### PR TITLE
[DNM] perform library reload after superfluous package load/custom node executions.

### PR DESCRIPTION
**Pull request by mjkkirschner**
### Purpose

@aparajit-pratap @reddyashish -

update:
current commit uses a pre-existing lock in the engine, to block reloading the library while the engine is inside a few important functions. One of these functions is `generateGraphSyncDataForCustomNode` https://github.com/DynamoDS/Dynamo/blob/dfb938bba667e9a4ddf54df1c983f3d4a48fdc1f/src/DynamoCore/Engine/EngineController.cs#L327

which is triggered when the customNode definition is updated. - this seems to stop the crash at least. But diagramming the execution flow - doesn't really explain why it works -  as I thought the graphSync generation would occur on the same thread as the library reload. (main thread)

`getMirror` also uses this lock
https://github.com/DynamoDS/Dynamo/blob/dfb938bba667e9a4ddf54df1c983f3d4a48fdc1f/src/DynamoCore/Engine/EngineController.cs#L212
and this might be a more likely explanation, as the graph executes getMirror maybe called from the scheduler thread and a race condition is prevented. - But I have not confirmed this.

In any case the reason this code works is not easy to confirm - and it could be that getting the lock simply changes timing enough to make avoid a race condition in debug mode. 😕 .

 But -  then we need to resync the custom node definitions with the engine, though I still don't understand exactly why this is case. My assumption is that reloading the library has dumped the loaded functions and they need to be re-registered - and usually this occurs after the library load as part of the node being loaded or placed. It's still not clear though or what consequences these changes have but they might be a potential refactor for later.

#### TLDR:

I'm not confident in this change at this time.

I found these two very interesting PRs:
https://github.com/DynamoDS/Dynamo/pull/2745
https://github.com/DynamoDS/Dynamo/pull/4235

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

**Comment by aparajit-pratap** [comment]
@mjkkirschner this will still cause the library loaded event handler to be scheduled after the graph execution. We need to make it happen before.

**Comment by mjkkirschner** [comment]
@aparajit-pratap - heres another solution.

this one works - but still the library is reloaded after the graph run - but they occur at different times because of the lock AND then all the custom nodes are resynced so they are available again in the VM.

So ideally yes we would reload the library before any node executions - this solution results in a bunch of wasted runs, lock contention (we could probably use the scheduler instead), and resyncing all the custom nodes twice....

I feel that this is way more risky to change than the small modification @reddyashish used in his pr to block the runs which cause the odd state in the first place.
